staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller_test.go

```diff
@@ -18,156 +18,291 @@ package controller
 
 import (
 	"context"
-	"io"
-	"os"
-	"path/filepath"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync/atomic"
 	"testing"
 	"time"
+
+	"k8s.io/apiserver/pkg/server/healthz"
+	"k8s.io/apiserver/pkg/server/options/encryptionconfig"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/component-base/metrics/legacyregistry"
+	"k8s.io/component-base/metrics/testutil"
 )
 
-func TestProcessEncryptionConfig(t *testing.T) {
-	testCases := []struct {
-		name        string
-		filePath    string
-		expectError bool
+func TestController(t *testing.T) {
+	tests := []struct {
+		name                     string
+		ecFilePath               string
+		apiServerID              string
+		validateECFileHash       bool
+		validateTransformerClose bool
+		validateSuccessMetric    bool
+		validateFailureMetric    bool
+		mockLoadEncryptionConfig func(ctx context.Context, filePath string, enableEncryption bool, apiServerID string) (*encryptionconfig.EncryptionConfiguration, error)
 	}{
 		{
-			name:        "empty config file",
-			filePath:    "testdata/empty_config.yaml",
-			expectError: true,
+			name:                     "when invalid config is provided previous config shouldn't be changed",
+			ecFilePath:               "testdata/empty_config.yaml",
+			apiServerID:              "test-apiserver",
+			validateECFileHash:       true,
+			validateTransformerClose: true,
+			validateFailureMetric:    true,
+			mockLoadEncryptionConfig: func(ctx context.Context, filePath string, enableEncryption bool, apiServerID string) (*encryptionconfig.EncryptionConfiguration, error) {
+				return &encryptionconfig.EncryptionConfiguration{
+					// hash of initial "testdata/ec_config.yaml" config file before reloading
+					EncryptionFileContentHash: "6bc9f4aa2e5587afbb96074e1809550cbc4de3cc3a35717dac8ff2800a147fd3",
+				}, fmt.Errorf("empty config file")
+			},
+		},
+		{
+			name:                  "when new valid config is provided it should be updated",
```

Comment:
@enj  waiting for 3 sec to check if `ctx / transformerClose` is "_not Done_" and passing the test is not possible since we always cancel parent context in the mock queue.